File: source/JAVA/iterator/latest/Hospital2.java
import java.time.Duration;
import java.time.Instant;
import java.util.Iterator;
import java.util.List;
import java.util.Objects;
import java.util.Random;

public class Hospital2 extends HospitalBase {

    static class Node implements Comparable<Node> {
        PatientBase data;
        Node next;

        Node(PatientBase patient) {
            data = patient;
        }

        @Override
        public int compareTo(Node o) {
            return data.compareTo(o.data);
        }

        @Override
        public String toString() {
            return "Node(data=" + data.toString() + ")";
        }
    }

    private Node head;

    @Override
    public boolean addPatient(PatientBase patient) {
        Node newNode = new Node(patient);
        Node curr = head;

        if (curr == null) {
            head = newNode;
            return true;
        }

        if (newNode.compareTo(curr) < 0) {
            newNode.next = curr;
            head = newNode;
            return true;
        }

        Node prev = null;
        while (newNode.compareTo(curr) >= 0) {
            if (curr.next == null) {
                curr.next = newNode;
                return true;
            }
            prev = curr;
            curr = curr.next;
        }

        if (prev != null) {
            prev.next = newNode;
        }
        newNode.next = curr;
        return true;
    }

    @Override
    public Iterator<PatientBase> iterator() {
        return new Iterator<>() {
            Node current = head;

            @Override
            public boolean hasNext() {
                return current != null;
            }

            @Override
            public PatientBase next() {
                if (hasNext()) {
                    PatientBase nextData = current.data;
                    current = current.next;
                    return nextData;
                }
                return null;
            }
        };
    }

    public static void main(String[] args) {
        /*
         * Public tests
         */
        {
            var hospital = new Hospital2();
            var p1 = new Patient("Max", "11:00");
            var p2 = new Patient("Alex", "13:15");
            var p3 = new Patient("George", "14:00");
            hospital.addPatient(p1);
            hospital.addPatient(p2);
            hospital.addPatient(p3);
            var patients = new Patient[] {p1, p2, p3};
            int i = 0;
            for (var patient : hospital) {
                assert Objects.equals(patient, patients[i++]);
            }
        }

        /*
         * Hidden tests
         */
        // One element only
        {
            var hospital = new Hospital2();
            var p1 = new Patient("Alex", "13:15");
            hospital.addPatient(p1);
            for (var patient : hospital) {
                assert Objects.equals(patient, p1);
            }
        }

        // Normal sort
        {
            var hospital = new Hospital2();
            var p1 = new Patient("Alex", "13:15");
            var p2 = new Patient("Max", "11:00");
            var p3 = new Patient("George", "14:00");
            var p4 = new Patient("Lisa", "09:00");
            hospital.addPatient(p1);
            hospital.addPatient(p2);
            hospital.addPatient(p3);
            hospital.addPatient(p4);
            var patients = List.of(p4, p2, p1, p3);
            int i = 0;
            for (var patient : hospital) {
                assert Objects.equals(patient, patients.get(i++));
            }
        }

        // Same names
        {
            var hospital = new Hospital2();
            var p1 = new Patient("Alex", "13:15");
            var p2 = new Patient("Alex", "11:00");
            var p3 = new Patient("Alex", "14:00");
            var p4 = new Patient("Alex", "09:00");
            hospital.addPatient(p1);
            hospital.addPatient(p2);
            hospital.addPatient(p3);
            hospital.addPatient(p4);
            var patients = List.of(p4, p2, p1, p3);
            int i = 0;
            for (var patient : hospital) {
                assert Objects.equals(patient, patients.get(i++));
            }
        }

        // Same times, different names
        {
            var hospital = new Hospital2();
            var p1 = new Patient("Daniel", "13:15");
            var p2 = new Patient("Mark", "13:15");
            var p3 = new Patient("Ben", "13:15");
            var p4 = new Patient("Alex", "13:15");
            hospital.addPatient(p1);
            hospital.addPatient(p2);
            hospital.addPatient(p3);
            hospital.addPatient(p4);
            var patients = List.of(p1, p2, p3, p4);
            int i = 0;
            for (var patient : hospital) {
                assert Objects.equals(patient, patients.get(i++));
            }
        }

        // Same times, different names 2
        {
            var hospital = new Hospital2();
            var p1 = new Patient("Alexc", "13:15");
            var p2 = new Patient("Alexz", "13:15");
            var p3 = new Patient("Alexb", "13:15");
            var p4 = new Patient("Alexa", "13:15");
            hospital.addPatient(p1);
            hospital.addPatient(p2);
            hospital.addPatient(p3);
            hospital.addPatient(p4);
            var patients = List.of(p1, p2, p3, p4);
            int i = 0;
            for (var patient : hospital) {
                assert Objects.equals(patient, patients.get(i++));
            }
        }

        // Time limit test 1 (iterate only first element)
        {
            var hospital = new Hospital2();
            var random = new Random();
            for (int i = 0; i < 1000; i++) {
                String patientName = random.ints(8, 'a', 'z' + 1)
                        .collect(StringBuffer::new, StringBuffer::append, StringBuffer::append)
                        .toString();
                List<String> hours = List.of("09", "10", "11", "13", "14", "15", "16", "17");
                String hour = hours.get(random.nextInt(hours.size()));
                int minutes = random.nextInt(60);
                String patientTime = String.format("%s:%02d", hour, minutes);
                hospital.addPatient(new Patient(patientName, patientTime));
            }
            var t1 = Instant.now();
            for (var patient : hospital) {
                break;
            }
            var t2 = Instant.now();
            assert Duration.between(t1, t2).toNanos() < 50_000;
        }

        // Time limit test 1 (add to beginning)
        {
            var hospital = new Hospital2();
            var random = new Random();
            for (int i = 0; i < 1000; i++) {
                String patientName = random.ints(8, 'a', 'z' + 1)
                        .collect(StringBuffer::new, StringBuffer::append, StringBuffer::append)
                        .toString();
                List<String> hours = List.of("09", "10", "11", "13", "14", "15", "16", "17");
                String hour = hours.get(random.nextInt(hours.size()));
                int minutes = random.nextInt(60);
                String patientTime = String.format("%s:%02d", hour, minutes);
                hospital.addPatient(new Patient(patientName, patientTime));
            }
            var t1 = Instant.now();
            hospital.addPatient(new Patient("a", "08:00"));
            var t2 = Instant.now();
            assert Duration.between(t1, t2).toNanos() < 50_000;
        }
    }
}

File: source/Python/iterator/hospital_2.py

import random
import string
import time

from hospital_base import HospitalBase
from patient_solution import Patient


class Node:
    def __init__(self, patient: Patient):
        self.data = patient
        self.next = None

    def __lt__(self, other):
        return self.data < other.data

    def __le__(self, other):
        return self.data <= other.data

    def __repr__(self):
        return f"Node(data={self.data.__repr__()})"


class Hospital_2 (HospitalBase):
    """
        Requirements for Hospital 2:
            - add_patient in O(n)
            - __iter__ - as quickly as possible
    """

    def __init__(self):
        super().__init__()
        self.head = None

    def __iter__(self):
        """
        Just iterate through the elements - they are already sorted
        """
        curr = self.head
        while (curr != None):
            yield curr.data
            curr = curr.next

    def add_patient(self, patient: Patient):
        """
        Sort the elements here
        """
        curr = self.head
        newNode = Node(patient)
        if curr is None:
            # no samples in the queue
            self.head = newNode
            return True

        if newNode < curr:
            newNode.next = curr
            self.head = newNode
            return True

        #print("Values ", newNode.data, curr.data)
        #if newNode == curr:
        #    print("HERE", newNode, curr)
        prev = None
        while newNode >= curr:

            if curr.next is None:
                # we got to the end
                curr.next = newNode
                return True
            else:
                prev = curr
                curr = curr.next

        if prev is not None:
            prev.next = newNode
        newNode.next = curr
        return True

    # ==============================  Add any extra functions below   ==============================


if __name__ == "__main__":

    """  ===================  Public tests   ===================  """
    ll = Hospital_2()
    ll.add_patient(Patient("Max", "11:00"))
    ll.add_patient(Patient("Alex", "13:15"))
    ll.add_patient(Patient("George", "14:00"))
    list_of_patients = [Patient("Max", "11:00"), Patient("Alex", "13:15"), Patient("George", "14:00")]
    for i, el in enumerate(ll):
        assert el == list_of_patients[i]
    assert i == 2
    """ ============= Hidden Tests ============= """
    # ================== One element only ==================
    ll = Hospital_2()
    ll.add_patient(Patient("Alex", "13:15"))
    for el in ll:
        assert el == Patient("Alex", "13:15")

    # ================== Normal Sort ==================
    ll = Hospital_2()
    ll.add_patient(Patient("Alex", "13:15"))
    ll.add_patient(Patient("Max", "11:00"))
    ll.add_patient(Patient("George", "14:00"))
    ll.add_patient(Patient("Lisa", "09:00"))
    list_of_patients = [Patient("Lisa", "09:00"), Patient("Max", "11:00"),
                        Patient("Alex", "13:15"), Patient("George", "14:00")]
    for i, el in enumerate(ll):
        assert el == list_of_patients[i]

    # ================== Same names ==================
    ll = Hospital_2()
    ll.add_patient(Patient("Alex", "13:15"))
    ll.add_patient(Patient("Alex", "11:00"))
    ll.add_patient(Patient("Alex", "14:00"))
    ll.add_patient(Patient("Alex", "09:00"))
    list_of_patients = [Patient("Alex", "09:00"), Patient("Alex", "11:00"),
                        Patient("Alex", "13:15"), Patient("Alex", "14:00")]
    for i, el in enumerate(ll):
        assert el == list_of_patients[i]

    # ================== Same times, different names ==================
    ll = Hospital_2()
    ll.add_patient(Patient("Daniel", "13:15"))
    ll.add_patient(Patient("Mark", "13:15"))
    ll.add_patient(Patient("Ben", "13:15"))
    ll.add_patient(Patient("Alex", "13:15"))
    list_of_patients = [Patient("Daniel", "13:15"), Patient("Mark", "13:15"),
                        Patient("Ben", "13:15"), Patient("Alex", "13:15")]
    for i, el in enumerate(ll):
        #print(el,  list_of_patients[i])
        assert el == list_of_patients[i]

    # ================== Same times, different names 2 ==================
    ll = Hospital_2()
    ll.add_patient(Patient("Alexc", "13:15"))
    ll.add_patient(Patient("Alexz", "13:15"))
    ll.add_patient(Patient("Alexb", "13:15"))
    ll.add_patient(Patient("Alexa", "13:15"))
    list_of_patients = [Patient("Alexc", "13:15"), Patient("Alexz", "13:15"),
                        Patient("Alexb", "13:15"), Patient("Alexa", "13:15")]
    for i, el in enumerate(ll):
        assert el == list_of_patients[i]

    # ================== Time limit test 1==================

    ll = Hospital_2()
    for i in range(1000):
        patient_name = ''.join(random.choice(string.ascii_lowercase) for _ in range(8))
        hour = random.choice(["09", "10", "11", "13", "14", "15", "16", "17"])
        minutes = random.choice(range(60))
        patient_time = f"{hour}:{minutes:02}"
        ll.add_patient(Patient(patient_name, patient_time))
    t1 = time.time()
    ll.add_patient(Patient('ZZZZZZ', "18:00"))
    t2 = time.time()
    # print(t2-t1)
    t1 = time.time()
    for i in ll:
        # only the first element - should be quick
        # no sorting is done in here
        break
    t2 = time.time()
    assert t2-t1 < 0.00005

    # ================== Time limit test 2==================
    # Insertion sort - if the value is added to the beginning- it is also quick
    ll = Hospital_2()
    for i in range(1000):
        patient_name = ''.join(random.choice(string.ascii_lowercase) for _ in range(8))
        hour = random.choice(["09", "10", "11", "13", "14", "15", "16", "17"])
        minutes = random.choice(range(60))
        patient_time = f"{hour}:{minutes:02}"
        ll.add_patient(Patient(patient_name, patient_time))
    t1 = time.time()
    ll.add_patient(Patient('a', "08:00"))
    t2 = time.time()
    assert t2-t1 < 0.00005